packages/cloud-functions/src/exchangeRate/FirebasePriceUpdater.ts

```diff
@@ -23,17 +21,30 @@ export default class FirebasePriceUpdater {
     this.manager = manager
   }
 
-  async refreshAllPrices(): Promise<PriceByAddress> {
+  async updateAllPrices(): Promise<PriceByAddress> {
     const prices = await this.manager.calculatecUSDPrices()
     await this.updatePrices(prices)
     return prices
   }
 
   private async updatePrices(prices: PriceByAddress) {
+    const fetchTime = Date.now()
+
     const tokensInfoRaw = await fetchFromFirebase(FIREBASE_NODE_KEY)
 
-    const fetchTime = Date.now()
+    await this.updatePresentTokens(prices, tokensInfoRaw, fetchTime)
 
+    const tokenAddresses = Object.values(tokensInfoRaw).map((tokenInfo: any) =>
+      tokenInfo.address?.toLowerCase()
+    )
+    const notAddedTokens = Object.keys(prices).filter(
+      (tokenAddress: string) => !tokenAddresses.includes(tokenAddress)
+    )
+
+    await this.updateNotPresentTokens(prices, notAddedTokens, fetchTime)
+  }
+
+  private async updatePresentTokens(prices: PriceByAddress, tokensInfoRaw: any, fetchTime: number) {
```

Comment:
Separate to my comment above ^, I'm not sure the `updatePresentTokens` has the behavior we want.  For example, token icon URLs can change and have, but we don't account for that.  E.g., relatively recent PR updating image URLs https://github.com/Ubeswap/default-token-list/pull/40/files

This maybe supports my suggestion above: if we need the complexity for some important reason, let's keep it.  Otherwise we will be better off w/ a simpler implementation that is easier to reason about.

Reply:
If we already have the information of a token, we don't need to use the token contract to get that information. So to avoid doing those calls, we have different behaviors. I can refactor an have a single `updateTokens` method, but that method will have some kind of `if (tokensPresents.include(token))` and we will split the behavior in it. Unless we want to update all the fields while fetching prices, which I think it's not needed.

Regarding the images, I think overriding the imageUrl from the Ubeswap list can be a bit limiting. For example, if in the future we want to manually change the imageUrl, we can't because this code will update the imageUrl again to the one that Ubeswap wants. So, I prefer to upload the image when adding all the token info, but not updating current imageUrls (kinda like a default value).
If we agree to use Ubeswap token list as our only source of truth, then we could start updating ours images automatically, but I'm not sure if we agree on that.
I have discussed this with Gonza, and we agree to use Ubeswap token list at the moment, but changing that on #1500.

Reply:
It is not a current requirement to allow updating image URLs manually. In #1500 we will host the images in our own infra, but we'll still be fetching the images from Ubeswap. I'm in favor of updating the images with whatever Ubeswap has.

As to your point about this being an optimization Silas, this is currently being called every 1 minute but may be called more often in the future. Updating all the tokens would mean making `3 * # of tokens` extra requests. If there is a large number of tokens (there are around 50 right now) this can get out of hand I think. My intuition is that it wouldn't be a problem _right now_ but it might become one in the future, so I'd favor leaving the actual implementation as is.

Reply:
Ok, I will change the code so it always checks Ubeswap list to update the image :)

Reply:
I'm not understanding the _specific_ impact of the optimization.  This is running as a cloud function, correct?  Why are we concerned with the number of network calls?  How specifically does that impact the user experience (or some other thing we should care about)?

Reply:
It takes time to make all those network requests (so our users see slightly more outdated prices), it spams the node we're using and we might incur bandwidth costs. There's no upper limit to the number of tokens and we're calling this every one minute now, so if there are hundreds or thousands of tokens will be making a very large number of requests every minute without adding any value. Short term we would probably run into Forno's rate limiting (we can ask them for an API key or run our own node to get around this).

Doing some quick math, lets say there are 1000 tokens (10x what we have today):
@ 3 requests per token we'd be making 3000 req/min = 50 requests per second.

That looks very high to me for something that doesn't really need it. With the optimization that number goes to 0 requests on average (each token is only fetched once, so average is practically zero).

I agree this optimization adds some complexity but I don't think is a very high amount of it. It's all self-contained to one file, maybe there's some code golf we could play to make it slightly simpler but I don't think avoiding the complexity is worth making so many requests. It would make everything slower and possibly bring us headaches in the future.

Reply:
:+1: 

The API limit sounds important.

It makes sense we want to avoid superfluous contract calls as a best practice.